### DISCUSSION TITLE
Why doesn't BusinessBase implement INotifyDataErrorInfo?

### DISCUSSION THREAD

**Discussion by GreatBarrier86** [MarimerLLC/csla#2382] [Questions]
I was curious about why ```BusinessBase``` doesn't implement ```INotifyDataErrorInfo```. I'm running into an issue with using CSLA with the DevExpress Xamarin control set and they've [indicated](https://supportcenter.devexpress.com/ticket/details/t1003055/follow-up-issue-to-t964238-dataformview-validation/) that the reason errors aren't showing up in certain situations is because ```INotifyDataErrorInfo``` is not being implemented. I can implement it myself, but I'm curious why it is not already done. I can see in [IBusinessBase](https://github.com/MarimerLLC/csla/blob/d9ea8628f394f976d0744be03fba23f701b546b2/Source/Csla/IBusinessBase.cs) that it's implemented when ```#if NETFX_CORE && !ANDROID && !IOS```, but in the scenario of Xamarin, it's being excluded. 

I could implement it like this which would be relatively simple, but it may be inappropriate
```
protected override void OnPropertyChanged(string propertyName)
        {
            base.OnPropertyChanged(propertyName);

            ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
        }

        public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;

        public bool HasErrors => BrokenRulesCollection.ErrorCount > 0;

        public IEnumerable GetErrors(string propertyName)
        {
            return BrokenRulesCollection.Where(p => p.Property == propertyName && p.Severity == RuleSeverity.Error).Select(p => p.Description);
        }
```

Any insight would be appreciated.

**Comment by rockfordlhotka** [comment]
There's also the older `IDataErrorInfo`, and you can only implement one or the other, because _some_ UI frameworks (WPF?) try to honor both and that causes issues.

In CSLA 6 (having just checked), I see that [`IDataErrorInfo` is implemented](https://github.com/MarimerLLC/csla/blob/1b9347e559ce5164a1ec9c024707851c499b4733/Source/Csla/Core/BusinessBase.cs#L53). This is actually a conundrum now that you mention it. Certainly CSLA 6 should implement `INotifyDataErrorInfo`, but that will almost certainly be problematic for WinForms users.

**Reply by GreatBarrier86** [reply]
Rocky, thanks for the info. Can you tell me how this would/could break WinForms? Our main app is still in that, so I may need to wrap this in compiler directives in order to prevent the WinForms app from breaking.

**Reply by rockfordlhotka** [reply]
Right, this needs some serious thought and consideration, and the problem isn't _new_ exactly.

It has been some time since we changed Csla.dll to target netstandard2.0, and it was at that point where only `IDataErrorInfo` became implemented. None of those other targets have been active for a couple years. I've removed all (most?) of them in the current codebase - but that was just cleanup of legacy code.

So WinForms pre-dates `INotifyDataErrorInfo` and so doesn't know anything about it. WinForms therefore relies entirely on `IDataErrorInfo` because that's what existed back then.

iirc, `INotifyDataErrorInfo` came into being around Silverlight, and so in the middle of WPF's lifespan. My recollection is that WPF got confused if both were implemented.

I strongly suspect _newer_ UI frameworks only look for `INotifyDataErrorInfo` and might ignore `IDataErrorInfo`, but that would need some testing.

In any case, because there are no longer Csla.dll builds _per UI framework_, this is problematic. Only one or the other can be implemented, unless I return to building the Csla project for the entire matrix of possible targets 😲

Though actually (train of consciousness here), it _might_ be that only WinForms is the outlier. Unfortunately I'm not sure WinForms is a NuGet target though. Hmm... 😕

**Reply by rockfordlhotka** [reply]
I wonder if I could _implement_ both, and have a config switch to enable/disable _behavior_ of one or the other?

**Reply by GreatBarrier86** [reply]
My thought was a switch as well, but I imagine it MAY pose a problem if the client code implements certain logic based on whether or not it seems the interface implemented. Maybe that’s not a problem in the large scheme of things but it’s basically “implementing it without implementing it”, if you will.

**Reply by hurcane** [reply]
We have a huge ERP solution with business objects that are shared by a Winforms client app and also Android mobile devices with Xamarin. I think the idea of implementing both with switches is a good plan. We don't use WPF in any of our projects, so I am not sure what problems would exist there. I am presuming the switches would control whether the interface events are raised?

**Reply by rockfordlhotka** [reply]
@hurcane that'll need to be tested though, because we need to confirm that _every_ UI framework (webforms, winforms, mvc, blazor, wpf, uwp, xamarin, maui, etc) can handle this concept.

My thought is that I can create some test code and maybe some of you can help me implement that code into some of these UI frameworks to spread the effort of this testing?

**Reply by GreatBarrier86** [reply]
Sorry. I'm still a little lost.  What issues would it cause with WinForms to implement both? If WinForms doesn't know anything about it, shouldn't it really not affect anything?

**Reply by rockfordlhotka** [reply]
@GreatBarrier86 I don't think it'll harm winforms. But it might break WPF, UWP, and other UI frameworks.

**Reply by GreatBarrier86** [reply]
oh, so when you said this, you meant if you removed IDataErrorInfo? I'm not trying to be pedantic. I'm just trying to catch up: "Certainly CSLA 6 should implement INotifyDataErrorInfo, but that will almost certainly be problematic for WinForms users."

I could offer help for WinForms and Xamarin. Eventually our Xamarin app will move to MAUI so I could also offer assistance there.

**Reply by rockfordlhotka** [reply]
@GreatBarrier86 no - my latest thought (that may or may not work) is to _implement both interfaces_, but have a config switch that makes one or the other "idle".

I'm pretty sure (not 100%) that this will work in WinForms, WebForms, and probably Blazor. I'm less convinced that it'll work for MVC, WPF, Xamarin, or UWP - all of which have roots during the time when they could have picked either (or both) to use.

**Reply by rockfordlhotka** [reply]
And to be clear: CSLA 6 _currently_ only implements `IDataErrorInfo` - which I didn't really think about until this thread. And that is clearly not acceptable.

**Reply by rockfordlhotka** [reply]
OK @GreatBarrier86  and @hurcane and anyone else who can help.

Here is a [simple repo](https://github.com/rockfordlhotka/DataErrorInfo) with the basis necessary to do a test.

It should be pretty easy to create a `PersonEdit` object, bind it to a UI, and see what does/doesn't work with the `EnableIDataErrorInfo` and `EnableINotifyDataErrorInfo` properties set to true/false in different combinations.

**Reply by GreatBarrier86** [reply]
Since i'm trying to accomplish the same task in our main Xamarin app, I thought I would implement the tests there as well, but I ran into a problem. I copied the interface implementation from your repository and tweaked a few things so it would work in my app, but the same logic applies. I tried putting a ```[field:NonSerialized]``` attribute on the event, but it had no effect. I then converted the event to have a backing field and applied the regular ```[NonSerialized]``` attribute to it, but that didn't work as well. I imagine this could possibly be a problem with a more complex implementation of your sample repo as well.

**During ```BeginEdit()```, the following error occurs:**
`System.NotSupportedException: 'Serialization of type object is not supported. (System.EventHandler``1[[System.ComponentModel.DataErrorsChangedEventArgs, System, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]])'`

```
        protected override void OnPropertyChanged(string propertyName)
        {
            base.OnPropertyChanged(propertyName);

            ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
        }

        private bool HasErrors()
        {
            return BrokenRulesCollection.ErrorCount > 0;
        }

        bool INotifyDataErrorInfo.HasErrors
        {
            get
            {
                return HasErrors();
            }
        }

        public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;

        IEnumerable INotifyDataErrorInfo.GetErrors(string propertyName)
        {
            return BrokenRulesCollection.Where(p => p.Property == propertyName);
        }
```

**Reply by rockfordlhotka** [reply]
@GreatBarrier86 I was trying to isolate the question from the context of CSLA - I see you are trying to do a test within the context of CSLA, which (on Xamarin) means these types need to subclass `Csla.Core.MobileObject` to participate in serialization.

**Reply by GreatBarrier86** [reply]
OK then i'll do so it without CSLA. Would be good to get some insight on the issue within the context of CSLA as well, if possible.

**Reply by rockfordlhotka** [reply]
We can do that if/when we establish that implementing both interfaces doesn't break the UI frameworks.

**Reply by GreatBarrier86** [reply]
That's fine. Question. I assume the process for testing this is for me to clone the git repo, add a Xamarin and/or WinForms project and implement the tests, and then do a Pull Request?

**Reply by rockfordlhotka** [reply]
I think that's a good plan.

**Reply by GreatBarrier86** [reply]
@rockfordlhotka, I'm unable to reference that project from a stock Visual Studio generated Xamarin project.  I tried with Standard 2.0 and 2.1

```
Error	NU1201	Project DataErrorInfo is not compatible with netstandard2.1 (.NETStandard,Version=v2.1). Project DataErrorInfo supports: net5.0 (.NETCoreApp,Version=v5.0)	DataErrorInfo.Xamarin
	
Error		Project '..\..\DataErrorInfo\DataErrorInfo.csproj' targets 'net5.0'. It cannot be referenced by a project that targets '.NETStandard,Version=v2.1'.	DataErrorInfo.Xamarin	C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets	1717	
```

Also, I know it's not related to this issue specifically but our main xamarin app needs to use INotifyDataErrorInfo and CSLA, so if you have the opportunity to look into the serialization thing, I'd appreciate it. Regardless of this test project, the control set we use expects that interface we're kinda stuck with that error.

**Reply by rockfordlhotka** [reply]
You should be able to copy `ErrorBase.cs` and `PersonEdit.cs` into your Xamarin project and use them.

**Reply by rockfordlhotka** [reply]
However, I did just update the repo so those classes are in a ns2.1 class library - that should make it easy.

**Reply by GreatBarrier86** [reply]
I'm pretty new to using Git since my company is currently using TFVC, but shouldn't there be other files here? Like your PersonEdit class and Library project?
![image](https://user-images.githubusercontent.com/29482393/127485533-72496d6a-4dbe-46b6-86b0-45e7016b761b.png)

**Reply by rockfordlhotka** [reply]
😳😳😳

Fixed. There was apparently an error in some biological system that was interacting with the keyboard.

**Reply by rockfordlhotka** [reply]
I did include a test for WPF, which appears to work fine with both turned on.

**Reply by GreatBarrier86** [reply]
I plan on doing the Xamarin and WinForms ones in the morning. Would be great to hammer this out.
